curvefs/src/mds/dlock/dlock.cpp

```diff
@@ -0,0 +1,214 @@
+/*
+ *  Copyright (c) 2021 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * Project: Curve
+ * Created Date: 2022-05-10
+ * Author: Jingli Chen (Wine93)
+ */
+
+#include <memory>
+#include <thread>
+
+#include <ostream>
+#include <functional>
+
+#include "src/common/encode.h"
+#include "src/common/timeutility.h"
+#include "curvefs/proto/mds.pb.h"
+#include "curvefs/src/mds/dlock/dlock.h"
+#include "curvefs/src/mds/common/storage_key.h"
+
+namespace curvefs {
+namespace mds {
+namespace dlock {
+
+using ::curve::common::TimeUtility;
+using ::curve::common::NameLockGuard;
+using ::curve::common::EncodeBigEndian;
+using ::curvefs::mds::DLockValue;
+using ::curvefs::mds::DLOCK_KEY_PREFIX;
+using ::curvefs::mds::DLOCK_PREFIX_LENGTH;
+
+std::ostream& operator<<(std::ostream& os,  OWNER_STATUS status) {
+    switch (status) {
+        case OWNER_STATUS::OK:
+            os << "OK";
+            break;
+        case OWNER_STATUS::ERROR:
+            os << "ERROR";
+            break;
+        case OWNER_STATUS::NOT_EXIST:
+            os << "NOT_EXIST";
+            break;
+        case OWNER_STATUS::EXPIRED:
+            os << "EXPIRED";
+            break;
+        default:
+            os << "UNKNOWN";
+            break;
+    }
+    return os;
+}
+
+DLock::DLock(const DLockOptions& options,
+             std::shared_ptr<EtcdClientImp> etcdClient)
+    : options_(options),
+      etcdClient_(etcdClient) {}
+
+LOCK_STATUS DLock::Lock(const std::string& name,
+                        const std::string& owner) {
+    bool succ = false;
+    uint64_t total = options_.tryTimeoutMs;
+    uint64_t step = options_.tryIntervalMs;
+    std::string key = EncodeKey(name);
+    LOG(INFO) << "DLock try lock, name=" << name << ", owner=" << owner
+              << "total=" << total << ", step=" << step << ", key=" << key;
+    for (uint64_t elapsed = 0; elapsed < total; elapsed += step) {
+        {
+            NameLockGuard lg(nameLock_, name);
+            if (ReplaceOwner(name, owner, options_.ttlMs)) {
+                succ = true;
+                break;
+            }
+        }
+        std::this_thread::sleep_for(std::chrono::milliseconds(step));
```

Comment:
you should use `bthread_usleep`, because the caller is within a bthread.

Reply:
> you should use `bthread_usleep`, because the caller is within a bthread.

fixed.